Auth with Directus and iOS blog post (#137)
directus-labs/guest-authoring: copy-this-template-directory/index.md

---

> title: 'Auth with Directus and iOS'
> description: 'Directus auth for an iOS project'

Reviewer: ```suggestion
title: 'Learn how to implement user registration and login in iOS with SwiftUI. Also discover how to lock down publishing and editing content with advanced user access control.'
```

author:
  name: 'Harshpal Bhirth'
  avatar_file_name: 'Harshpal.jpeg'
---

## Introduction

> In this tutorial, you will learn how to configure an iOS project that authenticates with Directus. You'll cover registering, logging in, logging out, viewing all posts from all users, creating a post, and editing and deleting posts from your user account.

Reviewer: ```suggestion
In this tutorial, you will learn how to configure an iOS project with Directus Auth. You'll cover registering, logging in, logging out, viewing all posts from all users, creating a post, and editing and deleting posts from your user account.
```

## Before You Start

You will need:

1. To have Xcode installed on your macOS machine.
2. Knowledge of the Swift programming language.
3. A Directus project - follow our [quickstart guide](docs.directus.io/getting-started/quickstart) if you don't already have one.

> ## Setting Up Public Role

Reviewer: ```suggestion
## Allowing Public Role To Create Users
```

1. Navigate to `Settings`.
2. Select `Access Control`.
3. Click on the `Public Role`.

> 4. Access the `System Collection`.
> 5. Locate `Directus_users`.
> 6. Under the `Create` option, choose `Use Custom`.
> 7. Enable field permissions for `Email` and `Password` options.

Reviewer: ```suggestion
4. Access the `System Collection`.
5. Locate `directus_users`.
6. Under the `Create` option, choose `Use Custom`.
7. Enable field permissions for `Email` and `Password` options.
```

Reviewer: Add screenshot too, please, of the field permissions screen with just email and password selected.

> ## Creating A New Role 

Reviewer: ```suggestion
## Creating A New iOS App User Role 
```

1. On the Access Control page, click the plus button.
2. Name your new role, such as `iOS App User`.
3. Under the "Post" collection, enable `Create` and `Read`.

> 4. For `Write` and `DELETE`, select `Use Custom`.

Reviewer: ```suggestion
4. For `Edit` and `Delete`, select `Use Custom`.
```

5. Add Filter `user_created -> id Equals $CURRENT_USER.id` for both update and delete options.
6. This configuration ensures users can create and read all posts, but only update and delete their own posts.
7. Copy the Primary Key from the iOS App User role.

> 8. In the Public Role settings, under `Directus_users`, select `Use Custom` for the create option.

Reviewer: ```suggestion
8. In the Public Role settings, under `Directus_users`, select `Use Custom` for the create option.
```

> 9. Click `Field Presets` and paste the Primary Key.

Reviewer: These needs rephrasing as it's unclear what to do. Also, add a screenshot here of the field preset being filled in.

> By following these steps, you'll set up a role in Directus that grants users of your iOS app the ability to view all posts but restricts editing and deleting permissions to posts associated with their authenticated user account.

Reviewer: Remove this, you've already said it above. BUT I would add that the users created by the public role will be given the iOS App User role.

> ## Content View

Reviewer: You're missing a bit of text under this that explains what we'll be doing / what to do with this code

Reviewer: Lets also add a screenshot here to make it easier to follow.

        
        let body: [String: Any] = [
            "refresh_token": refreshToken
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            print("Error encoding request body: \(error.localizedDescription)")
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                print("Error logging out: \(error.localizedDescription)")
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                if (200..<300).contains(httpResponse.statusCode) {

                    DispatchQueue.main.async {
                        isLoggedIn = false
                        accessToken = nil
                    }
                } else {
                    print("Failed to logout. Status code: \(httpResponse.statusCode)")
                }
            }
        }.resume()
    }
    
    struct ContentView_Previews: PreviewProvider {
        static var previews: some View {
            ContentView()
        }
    }
}

```
### ContentView Struct

> Defines the main view of the application.

Reviewer: Not required. Readers should already understand what this is.

### State Properties:

Three `@State` properties are declared to manage the state of the view:

- **showLoginView**: Tracks whether the login view should be displayed.
- **isLoggedIn**: Tracks whether the user is logged in.
- **accessToken**: Stores the access token after successful login.

### Body Property:

The `body` property defines the view's layout and behavior.

- **NavigationView**: Wraps the content and provides navigation functionality.
- **VStack**: Vertical stack layout for arranging child views.

> - **Spacer**: Flexible space to push content to the top and bottom.

Reviewer: Irrelevant. You don't need to explain every line - just those important to the functionality.

### Conditional Rendering:

Depending on the `isLoggedIn` state:
- If logged in, it displays navigation links for creating posts and viewing posts, along with a logout button.
- If not logged in, it displays buttons to login and register.

### Create Post NavigationLink:

Navigates to the `CreatePostView` when tapped.

### Posts NavigationLink:

> Navigates to the `PostsView` when tapped.

Reviewer: Both not required.

### Logout Button:

Triggers the `logout()` function when tapped.

### Logout Function:

- When the logout button is tapped, this function is called.
- It retrieves the `refreshToken` from the `accessToken` state variable.
- Constructs a POST request to the `/auth/logout` endpoint with the `refresh_token` included in the request body.
- If the logout request is successful (status code between 200 and 299), it updates the `isLoggedIn` state to `false` and clears the `accessToken`.
- If the logout request fails, it prints an error message with the status code.

> ## UserRegisterView

Reviewer: Again, explainer and screenshot

```swift
import SwiftUI

struct UserRegisterView: View {
    @Binding var isActive: Bool
    @State private var email: String = ""
    @State private var password: String = ""
    @State private var showAlert: Bool = false
    @State private var alertMessage: String = ""
    @Environment(\.presentationMode) var presentationMode
    
    var body: some View {
        VStack {
            TextField("Email", text: $email)
                .padding()
                .textFieldStyle(RoundedBorderTextFieldStyle())
            SecureField("Password", text: $password)
                .padding()
                .textFieldStyle(RoundedBorderTextFieldStyle())
            
            Button("Register") {
                registerUser()
            }
            .padding()
            .alert(isPresented: $showAlert) {
                Alert(title: Text("Error"), message: Text(alertMessage), dismissButton: .default(Text("OK")))
            }
        }
        .padding()
    }
    
    func registerUser() {
        guard !email.isEmpty, !password.isEmpty else {
            alertMessage = "Please enter both email and password"
            showAlert = true
            return
        }
        

>         guard let url = URL(string: "https://directus.lws.io/users") else {

Reviewer: ```suggestion
        guard let url = URL(string: "https://your-directus-project-url/users") else {
```

update throughout please

        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                showAlert(message: error.localizedDescription)
                return
            }
            
            if let data = data {
                if let responseString = String(data: data, encoding: .utf8) {
                    print("Response: \(responseString)")
                    DispatchQueue.main.async {
                        presentationMode.wrappedValue.dismiss()
                    }
                } else {
                    showAlert(message: "Failed to parse response")
                }
            } else {
                showAlert(message: "No data received")
            }
        }.resume()
    }
    
    func showAlert(message: String) {
        alertMessage = message
        showAlert = true
    }
}

struct UserRegisterView_Previews: PreviewProvider {
    static var previews: some View {
        UserRegisterView(isActive: .constant(false))
    }
}
```
### UserRegisterView Struct:

> Defines a view for user registration.

Reviewer: Not needed

Reviewer: Also apply earlier feedback - if it's reasonable to expect a Swift dev to know what boilerplate functions/variables are for, skip explaining them. If they're stylistic, skip them.

- **`@Binding var isActive: Bool`**: This is a binding variable that determines whether the view is active or not. Changes in this view propagate back to the parent.
- **`@State private var email`: String = ""**: State variable to hold the user's email address.
- **`@State private var password`: String = ""**: State variable to hold the user's password.
- **`@State private var showAlert`: Bool = false**: State variable to control whether to show an alert.
- **`@State private var alertMessage`: String = ""**: State variable to hold the message to be displayed in the alert.
- **`@Environment(\.presentationMode)` var presentationMode**: Environment variable to access the presentation mode, which allows the view to dismiss itself.

### Body View:

- **VStack**: A vertical stack that arranges its child views in a vertical line.
- **TextField**: Text field for entering the user's email.
- **SecureField**: Secure text field for entering the user's password.
- **Button**: Button to register the user.
- **.alert(isPresented: $showAlert)**: Modifier to present an alert when showAlert is true.

### registerUser Function:

This function is called when the user taps the "Register" button.

> - It first checks if the email and password fields are not empty. If they are empty, it sets the alertMessage and shows the alert.

Reviewer: ```suggestion
- It first checks if the email and password fields are not empty. If they are empty, it sets the `alertMessage` and shows the alert.
```

- Then it constructs a URL for the user registration endpoint `/user`.
- It creates a dictionary body containing the email and password.
- Converts the body dictionary to JSON data.
- Constructs a POST request with the JSON data in the HTTP body.
- Executes the request asynchronously using URLSession.shared.dataTask.
- Handles the response or any error that occurs during the network request.

> - If successful, it dismisses the view using the presentationMode.

Reviewer: I'm not sure how much of this is needed. Is it possible to just say it sends a POST request to /user with a payload containing email/password.

### showAlert Function:

- This function sets the alertMessage and sets showAlert to true, triggering the display of the alert.

## Login
```swift
import SwiftUI

struct LoginData: Codable {

>     let access_token: String

Reviewer: A little confused. When mentioning the logout function you say you extract the refresh_token, but here you only store the accessToken string?

        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: loginData, options: [])
        } catch {
            showAlert = true
            alertMessage = "Error encoding login data"
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data, let httpResponse = response as? HTTPURLResponse, error == nil else {
                showAlert = true
                alertMessage = error?.localizedDescription ?? "Unknown error"
                return
            }
            
            if (200..<300).contains(httpResponse.statusCode) {
                // Successful login
                if let loginResponse = try? JSONDecoder().decode(LoginResponse.self, from: data) {
                    accessToken = loginResponse.data.access_token
                    isLoggedIn = true
                    presentationMode.wrappedValue.dismiss()
                }
            } else {
                // Failed to login
                showAlert = true
                alertMessage = "Failed to login"
            }
        }.resume()
    }
}

> ```

Reviewer: Same notes below - strip it back to just what's important.

### Body View:

- Sets up a `VStack` containing two `TextField` views for inputting the post's title and content, respectively.
- Also contains a `Button` titled "Create Post", which triggers the `createPost()` function when tapped.
- Applies a modifier to show an alert if `showAlert` is true.

### createPost Function:

- Validates the URL for the endpoint where posts are created `/items/posts`.
- Constructs the post data dictionary with title and content.
- Creates a POST request with JSON-encoded post data.
- Adds necessary headers, including the authorization header with the provided access token.
- Performs a data task to execute the request asynchronously.
- Checks for errors and response status.
- If the status code indicates success (between 200 and 299), it prints a success message.
- If the status code indicates a failure, it shows an alert with an appropriate error message.

## Token Manager 

``` swift 
import Foundation

struct TokenManager {
    static let accessTokenKey = "accessToken"

    static func saveToken(_ accessToken: String, expirationTime: Date) {

>         UserDefaults.standard.set(accessToken, forKey: accessTokenKey)

Reviewer: Also needs the refresh key?
